### ISSUE TITLE
Introduce a consistent deprecation methodology

### ISSUE THREAD

**Issue by dmlloyd** [quarkusio/quarkus#5665]
We need a consistent deprecation methodology.  I'm not talking about a policy for when to drop things that are deprecated - this issue is about how to actually deprecate things.

There are two major areas where this applies:
* APIs (classes and members)
* Configuration properties and groups

Both I think can be handled the same way for the most part.  I would propose the following rules for both:
* Use the `@Deprecated` annotation
* Specify the `since` property (note that this requires JDK 9 or later so we might have to wait for this; once JDK 11 becomes a minimum we should carefully update all existing deprecations with accurate `since` versions)
* Specify `forRemoval` if the item is going to go away some day (should always, or nearly always, be `true`, else there wouldn't be a reason for deprecation) (also JDK 9+)
* In the element JavaDoc, provide an explanation in a `@deprecated` JavaDoc tag which includes what to use in replacement, if anything.

Using this methodology we can make sure things are marked deprecated in a timely manner, and we don't have to change the methodology every time we change the policy around rolling off deprecated things.

References:
* https://github.com/quarkusio/quarkus/pull/5387#discussion_r348881008
* #5596
* (add as needed)

**Comment by Sanne** [comment]
Sounds good. I'm missing one aspect: warning users when they are using deprecated configuration properties.

Do you intend to have the configuration tooling to automatically generate some warning based on the `@Deprecated` annotations on the configuration objects?

**Comment by dmlloyd** [comment]
Yes, I think that would be the right way to do it.  We could even include text derived from the `@deprecated` doc tag as well, e.g. `The property "foo.bar.baz" is deprecated: Please use "foo.blah.baz" instead.`  It should be possible to translate content of `{@link SomeConfig#someField}` to a property name string so that the JavaDoc can reference the field but the config message would reference the config property name.

**Comment by mkouba** [comment]
Just to clarify - does `APIs (classes and members)` only refer to the content of the `runtime` package from the runtime module? AFAIK there is no "API" package for deployment modules and so all public classes and members (build items, util classes, etc.) from the module can be considered part of the API.

**Comment by dmlloyd** [comment]
> Just to clarify - does `APIs (classes and members)` only refer to the content of the `runtime` package from the runtime module?

For the purposes of this issue, it applies to all Java code in the project.  When we talk about removal policy, that's when we'll have to decide what areas of code count as public API and what areas are internal.

**Comment by mkouba** [comment]
Ok, I see.

**Comment by rsvoboda** [comment]
Created enhancement request for automated warning about deprecated configuration properties - https://github.com/quarkusio/quarkus/issues/15486.
